java-checks/src/main/java/org/sonar/java/checks/helpers/RegexTreeHelper.java

```diff
@@ -74,4 +79,29 @@ private static void addCurrentGrapheme(List<RegexCheck.RegexIssueLocation> resul
     }
   }
 
+  public static boolean canReachWithoutConsumingInput(AutomatonState start, AutomatonState goal) {
+    return canReachWithoutConsumingInput(start, goal, new HashSet<>());
+  }
+
+  private static boolean canReachWithoutConsumingInput(AutomatonState start, AutomatonState goal, Set<AutomatonState> visited) {
+    if (start == goal) {
+      return true;
+    }
+    if (visited.contains(start)) {
+      return false;
+    }
+    visited.add(start);
+    for (AutomatonState successor : start.successors()) {
+      // We don't generally consider elements behind backtracking edges to be 0-input reachable because what comes
+      // after the edge won't directly follow what's before the edge. However, we do consider the end-of-lookahead
+      // state itself reachable (but not any state behind it), so that we can check whether the end of the lookahead
+      // can be reached without input from a given place within the lookahead.
+      if ((successor.incomingTransitionType() == EPSILON && canReachWithoutConsumingInput(successor, goal, visited))
+        || (successor instanceof EndOfLookaroundState && successor == goal)) {
```

Comment:
Removed backtracking edges from the list of traversable edges and added the ||-exception instead, as well as the comment above.